### PR TITLE
dev-libs/tree-sitter: support for building cli tool

### PR DESCRIPTION
**Pull request by msva**
It would be pretty useful to add a way to install tree-sitter-cli from TS source. It is pretty useful cli tool to work with grammars.

CC: @00-matt @sarnex ?
// PR as result of a conversation on IRC to make all interested people to discuss it here

### CODE REVIEW

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+		https://github.com/${PN}/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+		ts-cli? ( $(cargo_crate_uris) )
+	"
+	KEYWORDS="~amd64"
+fi
+
+LICENSE="MIT"
+SLOT="0"
+
+IUSE="ts-cli"
```
Comment (thesamesam) [inline dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:136]: Did you use 'pkgcheck scan'? You haven't added a description of this to metadata.xml.

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+	"${FILESDIR}/${PN}-No-static-libs-gentoo.patch"
+)
+
+src_unpack() {
+	if [[ ${PV} == *9999* ]]; then
+		git-r3_src_unpack
+		use ts-cli && cargo_live_src_unpack
+	else
+		cargo_src_unpack
+		# ^ covers "default" too, so can be left that way irrelevant
```
Comment (thesamesam) [inline dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:153]: Just put a comment above saying that it calls default so it's fine to do it unconditionally.

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+	EGIT_REPO_URI="https://github.com/${PN}/${PN}"
+else
+	SRC_URI="
+		https://github.com/${PN}/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+		ts-cli? ( $(cargo_crate_uris) )
+	"
+	KEYWORDS="~amd64"
+fi
+
+LICENSE="MIT"
```
Comment (thesamesam) [inline dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:133]: This doesn't seem likely to be correct with all the crates. See what cargo-ebuild says.

Comment (msva) [reply dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild]: Fixed.

Am I right in assumption that it if `cargo license` outputs "A **OR** B **OR** C" it is okay to take the one of them that is already on the list and ignore others?

Comment (00-matt) [reply dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild]: You can represent that in the LICENSE variable with `LICENSE="|| (A B C)"`.

I'm not too familiar with the cargo eclass, but if `CARGO_OPTIONAL=1` causes the extra crates to be downloaded when `USE=ts-cli`, you could make the extra licenses conditional on that.

```bash
# Core library is MIT, some crates licensed differently.
USE="
MIT
ts-cli? (
  || ( Apache-2.0 MPL-2.0 )
  Foo
)
"
```

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild (hunk 1 above)
Comment (msva) [reply dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:136]: Not yet.

Actually, I just initially posted ebuilds (and eclass) here to discuss the principles of changes, not as ready-to-push thing, so I didn't yet run neither `repoman` nor `pkgscan`.

But I already ran it before force-pushing commit with fixes.
And for now there is only 

```
dev-libs/tree-sitter
  DroppedKeywords: version 0.20.1-r1: alpha, arm, arm64, hppa, ia64, ppc, ppc64, riscv, s390, sparc, x86
```

But I'm not sure if I should add them even as `~`, so I'll keep it as is.

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+	EGIT_REPO_URI="https://github.com/${PN}/${PN}"
+else
+	SRC_URI="
+		https://github.com/${PN}/${PN}/archive/${PV}.tar.gz -> ${P}.tar.gz
+		ts-cli? ( $(cargo_crate_uris) )
+	"
+	KEYWORDS="~amd64"
+fi
+
+LICENSE="Apache-2.0 BSD-2 CC0-1.0 ISC MIT"
```
Comment (sarnex) [inline dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:133]: If the non-MIT licenses are from the cli tool stuff, we should probably have a conditiona license

LICENSE="MIT ts-cli? ( Apache-2.0 BSD-2 CC0-1.0 ISC )" or whatever

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild (hunk 1 above)
Comment (sarnex) [reply dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:136]: Do we have a reason to drop all those keywords?

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild (hunk 4 above)
Comment (msva) [reply dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:133]: Yeah, we discussed that on IRC already.
I've already changed that locally, just we didn't find final conclusion about how that should look like to properly reflect the "OR"'s in crates licenses.

The thing is in that actually `cargo licanse` says that:

```
Apache-2.0 (1): chunked_transfer
Apache-2.0 OR Apache-2.0 WITH LLVM-exception OR MIT (2): wasi, wasi
Apache-2.0 OR BSL-1.0 (1): ryu
Apache-2.0 OR MIT (75): anyhow, arrayvec, ascii, autocfg, base64, bitflags, bumpalo, cc, cfg-if, chrono, crossbeam-utils, ctor, diff, dirs, dirs-sys, either, form_urlencoded, getrandom, getrandom, glob, hashbrown, hermit-abi, idna, indexmap, itoa, js-sys, lazy_static, libc, log, num-integer, num-traits, once_cell, percent-encoding, ppv-lite86, pretty_assertions, proc-macro2, quote, rand, rand_chacha, rand_core, rand_hc, regex, regex-syntax, remove_dir_all, rust-argon2, rustc-hash, serde, serde_derive, serde_json, smallbitvec, syn, tempfile, thiserror, thiserror-impl, thread_local, time, tiny_http, toml, unicode-bidi, unicode-normalization, unicode-width, unicode-xid, url, vec_map, wasm-bindgen, wasm-bindgen-backend, wasm-bindgen-macro, wasm-bindgen-macro-support, wasm-bindgen-shared, web-sys, webbrowser, widestring, winapi, winapi-i686-pc-windows-gnu, winapi-x86_64-pc-windows-gnu
Apache-2.0 OR MIT OR Zlib (2): tinyvec, tinyvec_macros
BSD-2-Clause (1): arrayref
CC0-1.0 (1): constant_time_eq
ISC (1): libloading
MIT (23): ansi_term, ansi_term, atty, blake2b_simd, clap, difference, html-escape, matches, output_vt100, redox_syscall, redox_syscall, redox_users, spin, strsim, textwrap, tree-sitter, tree-sitter-cli, tree-sitter-config, tree-sitter-highlight, tree-sitter-loader, tree-sitter-tags, utf8-width, which
MIT OR Unlicense (5): aho-corasick, memchr, same-file, walkdir, winapi-util
```

And all that I can imagine to reflect that is:
```
LICENSE="
        MIT
        ts-cli? (
                Apache-2.0
                || ( Apache-2.0 OR Apache-2.0-with-LLVM-exceptions MIT )
                || ( Apache-2.0 Boost )
                || ( Apache-2.0 MIT )
                || ( Apache-2.0 MIT ZLIB )
                BSD-2
                CC0-1.0
                ISC
                MIT
                || ( MIT Unlicense )
        )
"
```

But I'm unsure about such amount of duplication.

And I also unsure if
```
LICENSE="
  MIT
  ts-cli? ( Apache-2.0 BSD-2 CC0-1.0 ISC MIT )
"
```

would suffice...

And what do you think about it?

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild (hunk 1 above)
Comment (msva) [reply dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:136]: Oh. Sorry. I mssed a moment when did they disappeared.
I was totally sure I took `0.20.1` ebuild from gentoo repo and made the changes to it and didn't touch the KEYWORDS on my own.
But now I see keywords are there, so idk what's going on.

So, I'll merge KEYWORDS from there then (after we'll finish with LICENSES)

File: dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild (hunk 4 above)
Comment (sarnex) [reply dev-libs/tree-sitter/tree-sitter-0.20.1-r1.ebuild:133]: Looks correct to me

**Comment by msva** [comment]
btw, there is also a warning from `pkgcheck scan`:
```
  NonsolvableDepsInDev: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~arm) dev profile (default/linux/arm/17.0/armv4) (14 total): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~alpha) stable profile (default/linux/alpha/17.0) (16 total): solutions: [ virtual/rust ]
```

Can it be ignored since we have dependency on rust only with `ts-cli` flag (and KEYWORDS have no support to be use-conditional), or should keywords be dropped to have only common ones with `virtual/rust`?

**Comment by sarnex** [comment]
> btw, there is also a warning from `pkgcheck scan`:
> 
> ```
>   NonsolvableDepsInDev: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~arm) dev profile (default/linux/arm/17.0/armv4) (14 total): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~alpha) stable profile (default/linux/alpha/17.0) (16 total): solutions: [ virtual/rust ]
> ```
> 
> Can it be ignored since we have dependency on rust only with `ts-cli` flag (and KEYWORDS have no support to be use-conditional), or should keywords be dropped to have only common ones with `virtual/rust`?

Dropping alpha makes sense since rust isn't supported at all. arm should be fine though, i'm asking the rust team whats going on

**Comment by sarnex** [comment]
> btw, there is also a warning from `pkgcheck scan`:
> 
> ```
>   NonsolvableDepsInDev: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~arm) dev profile (default/linux/arm/17.0/armv4) (14 total): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~alpha) stable profile (default/linux/alpha/17.0) (16 total): solutions: [ virtual/rust ]
> ```
> 
> Can it be ignored since we have dependency on rust only with `ts-cli` flag (and KEYWORDS have no support to be use-conditional), or should keywords be dropped to have only common ones with `virtual/rust`?

nevermind, rust is only suppoed on some hardware supported by the arm keyword, lets drop it too

**Comment by msva** [comment]
also:

```
dev-libs/tree-sitter
  DroppedKeywords: version 0.20.1-r1: alpha, arm
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~hppa) stable profile (default/linux/hppa/17.0): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~hppa) stable profile (default/linux/hppa/17.0/desktop): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0/desktop): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0/desktop/gnome): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0/desktop/gnome/systemd): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/64ul): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/64ul/desktop): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/64ul/systemd): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/desktop): solutions: [ virtual/rust ]
  NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/systemd): solutions: [ virtual/rust ]
```

I think it is pretty safe drop them too, as there is almost no alive hw using them

**Comment by sarnex** [comment]
> also:
> 
> ```
> dev-libs/tree-sitter
>   DroppedKeywords: version 0.20.1-r1: alpha, arm
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~hppa) stable profile (default/linux/hppa/17.0): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~hppa) stable profile (default/linux/hppa/17.0/desktop): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0/desktop): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0/desktop/gnome): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~ia64) stable profile (default/linux/ia64/17.0/desktop/gnome/systemd): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/64ul): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/64ul/desktop): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/64ul/systemd): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/desktop): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~sparc) stable profile (default/linux/sparc/17.0/systemd): solutions: [ virtual/rust ]
> ```
> 
> I think it is pretty safe drop them too, as there is almost no alive hw using them

Yeah we need to fix the warnings, so I guess drop anything that warns and isn't fixable

**Comment by msva** [comment]
So:
```
dev-libs/tree-sitter
  DroppedKeywords: version 0.20.1-r1: alpha, arm, hppa, ia64, sparc
```

**Comment by sarnex** [comment]
> So:
> 
> ```
> dev-libs/tree-sitter
>   DroppedKeywords: version 0.20.1-r1: alpha, arm, hppa, ia64, sparc
> ```

lgtm

**Comment by thesamesam** [comment]
> 

just mask it on the profiles which don't work please, there's an existing package.mask you can tag onto.

anyway, if the dependency is optional, please package.use.mask the rust flags instead. pkgcheck needs this so we can't lose the keywords forever

**Comment by sarnex** [comment]
yeah sorry that's much better, we can keep the keywords but mask the use on unsupported profiles

**Comment by thesamesam** [comment]
> btw, there is also a warning from `pkgcheck scan`:
> 
> ```
>   NonsolvableDepsInDev: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~arm) dev profile (default/linux/arm/17.0/armv4) (14 total): solutions: [ virtual/rust ]
>   NonsolvableDepsInStable: version 0.20.1-r1: nonsolvable depset(bdepend) keyword(~alpha) stable profile (default/linux/alpha/17.0) (16 total): solutions: [ virtual/rust ]
> ```
> 
> Can it be ignored since we have dependency on rust only with `ts-cli` flag (and KEYWORDS have no support to be use-conditional), or should keywords be dropped to have only common ones with `virtual/rust`?

As covered in the quizzes, this is exactly why package.use.mask exists.

**Comment by msva** [comment]
> As covered in the quizzes, this is one of several reasons why package.use.mask exists.

Uh, shame on me, I didn't recall that while thinking about this situation.

**Comment by msva** [comment]
Done

File: profiles/arch/alpha/package.use.mask [hunk 5]
```diff
@@ -502,3 +502,7 @@ net-dialup/ppp atm
 
 # Needs nut and qstat respectively, neither is keyworded on alpha.
 net-analyzer/nagios-plugins nagios-game
+
+# Vadim Misbakh-Soloviov <mva@gentoo.org> (2021-12-09)
```
Comment (thesamesam) [inline profiles/arch/alpha/package.use.mask:506]: Please put new [entries](https://devmanual.gentoo.org/profiles/package.mask/index.html) at the _top_ of the mask files (below any headers).

**Review by sarnex — APPROVED** [review]

**Review by thesamesam — APPROVED** [review]

**Pull request closed without merging**